### PR TITLE
fix: ensure properly wrapping in pytorch. [DET-3745]

### PR DESCRIPTION
**Pull request by shiyuann**
## Description

Users might not refer to the wrapped optimizer when creating an LR scheduler. We used to have checks in `wrap_lr_scheduler` but not in the backward compatibility code with the old API. This PR fixes that and adds a debugging FAQ to the Pytorch trial documentation for those kind of issues.

## Test Plan

Use the CI test suite.

## Commentary (optional)

N/A

[DET-1234]: https://determinedai.atlassian.net/browse/DET-1234

### CODE REVIEW

File: harness/determined/pytorch/_pytorch_context.py [hunk 1]
```diff
@@ -189,12 +189,19 @@ def wrap_lr_scheduler(
...
-            lr_scheduler.optimizer,  # type: ignore
-            self.optimizers,
-            "Must use an optimizer that is returned by wrap_optimizer()",
-        )
+            )
+        else:
+            logging.warning(
+                "You are using custom LR scheduler, please remember to use the wrapped"
+                "optimizer returned by wrap_optimizer."
+            )
```
Comment (rb-determined-ai) [inline harness/determined/pytorch/_pytorch_context.py:204]: Can we avoid emitting "just in case" warnings when we don't know that something is wrong?

File: harness/determined/pytorch/_pytorch_trial.py [hunk 2]
```diff
@@ -171,6 +171,20 @@ def _apply_backwards_compatibility(self) -> None:
...
+                        self.context.optimizers,
+                        "Must use a wrapped optimizer that is passed in by the optimizer argument of "
+                        "create_lr_scheduler",
+                    )
+                else:
+                    logging.warning(
+                        "You are using custom LR scheduler, please remember to use a wrapped "
+                        "optimizer that is passed in by the optimizer argument of "
+                        "create_lr_scheduler."
+                    )
```
Comment (rb-determined-ai) [inline harness/determined/pytorch/_pytorch_trial.py:187]: ditto

File: docs/reference/api/pytorch.txt [hunk 3]
```diff
@@ -170,6 +170,18 @@ your code:
    to the ``clip_grads`` argument of ``step_optimizer``
    while removing the ``PyTorchCallback`` counterpart in the :meth:`build_callbacks` method.
 
+Debugging FAQ
+-------------
+
+* **Q: If my job got paused (e.g. if a higher weight job gets scheduled over the top of a running job),
+  then the resulting validation accuracies are significantly lower than for jobs that weren't paused,
+  even though the training accuracies are the same.**
```
Comment (rb-determined-ai) [inline docs/reference/api/pytorch.txt:178]: I know that we talked about adding an FAQ earlier, but now that I read this, I think this is way too specific.

Maybe we could rephrase the question to be something more generic, like:
```
Q: Why am I seeing strange metrics for trials which are paused and later continued (jumps in loss, worse convergence, etc)?
```

File: docs/reference/api/pytorch.txt [hunk 4]
```diff
@@ -170,6 +170,18 @@ your code:
    to the ``clip_grads`` argument of ``step_optimizer``
    while removing the ``PyTorchCallback`` counterpart in the :meth:`build_callbacks` method.
 
+Debugging FAQ
```
Comment (rb-determined-ai) [inline docs/reference/api/pytorch.txt:173]: Move this to a pytorch section of the main FAQ.

File: docs/reference/api/pytorch.txt [hunk 5]
```diff
@@ -170,6 +170,18 @@ your code:
...
+-------------
+
+* **Q: If my job got paused (e.g. if a higher weight job gets scheduled over the top of a running job),
+  then the resulting validation accuracies are significantly lower than for jobs that weren't paused,
+  even though the training accuracies are the same.**
+
+  * A: Please check if you have wrapped your model, optimizer, and LR scheduler with
```
Comment (rb-determined-ai) [inline docs/reference/api/pytorch.txt:180]: How about something a bit more organized, like:

```
  * A: Please verify the following:
          * your model is wrapped with ``wrap_model``
          * your optimizer is wrapped with ``wrap_optimizer`` and is based on output of ``wrap_model``, not the original unwrapped model
          * your LR scheduler is wrapped with ``wrap_lr_scheduler`` and is based on the output of ``wrap_optimizer``, not the original unwrapped optimizer
```

(you'll have to edit it for reST syntax and stuff)

File: docs/reference/api/pytorch.txt (hunk 4 above)
Comment (shiyuann) [reply docs/reference/api/pytorch.txt:173]: We don't have the main FAQ and I don't think putting everything in one main FAQ would be helpful to users. Basically we haven't discussed how we would have FAQ. I think it's fine to just put that here.

Comment (rb-determined-ai) [reply docs/reference/api/pytorch.txt:173]: Yes we do: https://docs.determined.ai/latest/faq.html, and it already has a "tensorflow support" section.

File: docs/reference/api/pytorch.txt [hunk 6]
```diff
@@ -170,6 +170,20 @@ your code:
...
+-------------
+
+* **Q: Why am I seeing significantly different metrics for trials which are paused and later
+  continued (jumps in loss, worse convergence, etc) than trials which aren't paused?**
+
+  * A: Please verify the following:
+
+    * your model is wrapped with :meth:`wrap_model <determined.pytorch.PyTorchTrialContext.wrap_model>`.
+    * your optimizer is wrapped with :meth:`wrap_optimizer <determined.pytorch.PyTorchTrialContext.wrap_optimizer>`
+      and is based on output of ``wrap_model``, not the original unwrapped model.
```
Comment (rb-determined-ai) [inline docs/reference/api/pytorch.txt:183]: "based on **the** output of"

File: docs/reference/api/pytorch.txt (hunk 4 above)
Comment (shiyuann) [reply docs/reference/api/pytorch.txt]: oops. I am surprised that was not linked to the framework-specific doc. Moved the doc to that main FAQ.

**Comment by shiyuann** [comment]
e2e-test: https://app.circleci.com/pipelines/github/determined-ai/determined/4911/workflows/766050bf-c9a1-4d4a-a7c6-23e20f3e0723
